feat(client): ✨ Sign out button/Sign in modal

Muaad2001 opened the pull request:
User is able to sign-out from there account. Signout button appears after user is signed in.
User Can create an account using Microsoft or Google
User can Create an account as well as Signing into that account
Sign in appears as a modal.

Muaad2001 committed:
Signout almost done

Muaad2001 committed:
sign out feature done

Muaad2001 committed:
feat(client): ✨ Sign in with email drop down

Sign in with email and password has drop down and is in the modal

Muaad2001 committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into signin-feature

Muaad2001 committed:
feat(client): 🔥 Moved code from signin file into modal, deleted code from sign in folder

Muaad2001 committed:
Merge branch 'Averbach-Abdulaziz-signin' of https://github.com/cat-acronym-team/cat-me-if-you-can into Muaad-Signout

Muaad2001 committed:
feat(client): ✨ User is able to sign out

Muaad2001 committed:
style(client): 🎨 Add Images for google and microsoft log in

Add image for google and Microsoft log in. Add X button to exit out of modal when signing in

nstringham reviewed (changes requested):
This PR clearly adds a lot more than just a sign out button. let's change the title to reflect that.

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,79 @@
+      </button>
+    </div>
+    <div class="email">
+      <button on:click={clickedEmailButton}>Sign In</button>
```
this should probobly say sign in with email and password or something like that

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,79 @@
+    if (email == "") {
+      return;
+    }
+    await loginWithEmail(email, password);
```
you should defintly be catching erros here

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,79 @@
+  let email = "";
+  let clickedButton = false;
+  async function submitLogin() {
+    if (password == "") {
+      return;
+    }
+    if (email == "") {
+      return;
+    }
```
should probobly also disable the button in this case

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -31,13 +91,27 @@
-        <a href="/stats">Stats</a>
-        <a href="/logout">Logout</a> -->
+        <a href="/stats">Stats</a> -->
+      <a href="" on:click={onSignOut}>Logout</a>
```
this should probobly be a button

nstringham commented:
> This PR clearly adds a lot more than just a sign out button. let's change the title to reflect that.

Oh I see. it got mixed up with #76

Muaad2001 committed:
fix(client): 🎨 Add button for sign out

Muaad2001 committed:
perf(client): ⚡️ Add error messages when user enter invalid email or password

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -35,3 +44,14 @@ export async function loginWithEmail() {
+// Sign out
+export function onSignOut() {
+  signOut(auth)
+    .then(() => {
+      // Sign-out successful.
+    })
+    .catch((error) => {
+      // An error happened.
+    });
```
if these are just going to be empty we may as well remove them

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -35,3 +44,14 @@ export async function loginWithEmail() {
+// Sign out
+export function onSignOut() {
+  signOut(auth)
+    .then(() => {
+      // Sign-out successful.
+    })
+    .catch((error) => {
+      // An error happened.
+    });
```
if these are just going to be empty we may as well remove them

Muaad2001 replied:
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,90 @@
+      try {
+        await loginWithEmail(email, password);
+      } catch (err) {
+        let errorMessage: string;
+        if (err instanceof Error) {
+          errorMessage = err.message;
+        } else {
+          errorMessage = String(err);
+        }
+        console.error("Username or Password is invalid", errorMessage);
```
this should be displayed to the user

Muaad2001 replied:
resolved

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -23,10 +31,11 @@ export async function loginWithMicrosoft() {
+    const user = await createUserWithEmailAndPassword(auth, email, password);
+    console.log(user);
   } catch (error) {
     console.log(error);
```
the error proboly shouldn't be cought here. personally I would just return createUserWithEmailAndPassword directly

Muaad2001 replied:
resolved

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -23,10 +31,11 @@ export async function loginWithMicrosoft() {
   try {
-    // const user = await createUserWithEmailAndPassword(auth, email, password);
-    // console.log(user);
+    const user = await createUserWithEmailAndPassword(auth, email, password);
```
does createUserWithEmailAndPassword work for returning users?

Muaad2001 replied:
resolved

Muaad2001 replied:
changed from createUser to signInWithEmailAndPassword

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,90 @@
+            </div>
+            <div class="formGroup">
+              <label for="password"><b>Password</b></label>
+              <input type="text" bind:value={password} placeholder="Enter Password" name="password" required />
```
maybe conside `type="password"`
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/password

Muaad2001 replied:
resolved

Muaad2001 committed:
feat(client): ✨ Add create account with email and password

Users have to create their account using email and password before signing in with email and password

Muaad2001 committed:
fix(client): 🎨 Error code is displayed to user when invalid username or password is entered

Muaad2001 committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into Muaad-Signout

Muaad2001 committed:
fix(client): 🐛 Reset errorMessage to "" after valid input is finally entered to close modal

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -23,15 +32,20 @@ export async function loginWithMicrosoft() {
+
+// Sign out
+export function onSignOut() {
+  signOut(auth);
```
this should probobly return

Muaad2001 replied:
resolved

nstringham commented on client/static/images/stuff2.svg:
```diff
@@ -0,0 +1,9 @@
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="256" height="256" viewBox="0 0 256 256" xml:space="preserve">
```
can we please give this file a more descriptive name

Muaad2001 replied:
oh yeah sorry about that lol, resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+  let email = "";
+  let password = "";
+  let confirmPass = "";
+  let clickedButton = false;
+  let clickedButton2 = false;
```
can we please give these variables more descriptive names

Muaad2001 replied:
clickedButton -> signInButton
clickedButton2 -> createAccountButton
Resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+  let clickedButton2 = false;
+  let errorMessage = "";
+  async function createAccount() {
+    if (password == "") {
+      return;
+    }
+    if (password != confirmPass) {
+      errorMessage = "Password does not match";
+      password = "";
```
the button should be disabled if the password is empty so that the use knows they cannot click it

Muaad2001 replied:
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+    if (password != confirmPass) {
+      errorMessage = "Password does not match";
+      password = "";
+      confirmPass = "";
+      return;
+    }
```
a message should be displayed to the user to explain why nothing happened when they clicked the button

Muaad2001 replied:
on line 19 i set errorMessage to "Password do not match" and I display that to the user, ill move that to the return statment
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+      confirmPass = "";
+      return;
+    }
+    if (email == "") {
+      return;
```
the button should be disabled if the email is empty so that the use knows they cannot click it

Muaad2001 replied:
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+    }
+    if (email == "") {
+      return;
+    } else {
```
this `else` is redundant because the `if` returns

Muaad2001 replied:
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+        errorMessage = "";
+      } catch (err) {
+        errorMessage = "Invalid Username or Password";
+      }
```
not all errors are a result of an Invalid Username or Password you should check the `err` object and display a relivant message to the user

Muaad2001 replied:
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+    if (errorMessage == "") {
+      openSignInModal = false;
+    }
+    email = "";
+    password = "";
+    confirmPass = "";
```
are we sure we want to reset these fields even when the signup was not successful?

Muaad2001 replied:
moved them into the else statement so they only reset if signup is successful
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+    confirmPass = "";
+  }
+  async function submitLogin() {
+    if (password == "") {
+      return;
+    }
+    if (email == "") {
+      return;
```
the button should be disabled if the password or email is empty so that the use knows they cannot click it

Muaad2001 replied:
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+    }
+    if (email == "") {
+      return;
+    } else
```
this `else` is redundant because the `if` returns

Muaad2001 replied:
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+        await loginWithEmail(email, password);
+        errorMessage = "";
+      } catch (err) {
+        errorMessage = "Invalid Username or Password";
```
not all errors are a result of an Invalid Username or Password you should check the `err` object and display a relivant message to the user

Muaad2001 replied:
changed to  errorMessage = err instanceof Error ? err.message : String(err);
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+    password = "";
+    email = "";
+  }
+  function signInWithEmailAndPassword() {
+    clickedButton = !clickedButton;
+    clickedButton2 = false;
+  }
+  function createWithEmailAndPassword() {
+    clickedButton2 = !clickedButton2;
+    clickedButton = false;
+  }
```
it might be better to give these functions names that more closly reflect there functionallity. they are toggeling the visiblity of forms not really signing in or creating an account

Muaad2001 replied:
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+    clickedButton = false;
+  }
+  async function googleLogin() {
+    await loginWithGoogle();
```
we should be catching and displaying errors here

Muaad2001 replied:
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
+  }
+
+  async function microsoftLogin() {
+    await loginWithMicrosoft();
```
we should be catching and displaying errors here

Muaad2001 replied:
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,151 @@
-  }}
->
+<Modal open={openSignInModal}>
   <!-- TODO: Sign In Modal Content Here -->
```
we probobly don't need this TODO anymore

Muaad2001 replied:
resolved

Muaad2001 committed:
refactor(client): 🔥 Deleted if statments from signin and createaccount functions

Lots of redundant if statements, Added return value for signout, ErrorMessage properly displays err

Muaad2001 committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into Muaad-Signout

gchamb commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,141 @@
+    if (password != confirmPass) {
+      password = "";
+      confirmPass = "";
+      return (errorMessage = "Passwords do not match");
```
You do not need to return this and you aren't using this return value anywhere. Just assign errorMessage to the message then just return.  Example: `errorMessage = "Wrong stuff"; return; `

gchamb commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,141 @@
+      await createUser(email, password);
+      errorMessage = "";
+    } catch (err) {
+      return (errorMessage = err instanceof Error ? err.message : String(err));
```
You do not need to return this and you aren't using this return value anywhere. Just assign errorMessage to the message

gchamb commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,141 @@
+      await loginWithGoogle();
+      errorMessage = "";
+    } catch (err) {
+      return (errorMessage = err instanceof Error ? err.message : String(err));
+    }
```
You do not need to return this and you aren't using this return value anywhere. Just assign errorMessage to the message

gchamb commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,141 @@
+      await loginWithMicrosoft();
+      errorMessage = "";
+    } catch (err) {
+      return (errorMessage = err instanceof Error ? err.message : String(err));
```
You do not need to return this and you aren't using this return value anywhere. Just assign errorMessage to the message

gchamb commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,141 @@
+    } catch (err) {
+      return (errorMessage = err instanceof Error ? err.message : String(err));
+    }
+    openSignInModal = false;
```
this closes the signin modal if  i exit out the google popup. do we want that?

gchamb commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,141 @@
+    } catch (err) {
+      return (errorMessage = err instanceof Error ? err.message : String(err));
+    }
+    openSignInModal = false;
```
this closes the signin modal if i exit out the microsoft popup. do we want that?

gchamb commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,141 @@
+          </div>
+        </div>
+        <div class="formButton">
+          <button type="submit">Create Account</button>
```
this button still shows if i click sign in with email and password

gchamb commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,141 @@
+      await loginWithEmail(email, password);
+      errorMessage = "";
+    } catch (err) {
+      return (errorMessage = err instanceof Error ? err.message : String(err));
```
do we want a better error message? The error message returns `Firebase: Error (auth/user-not-found).`

gchamb commented on client/src/lib/firebase/auth.ts:
```diff
@@ -1,12 +1,21 @@
+    console.log(user);
   } catch (error) {
     console.log(error);
   }
```
no need for try catch if you're try catching when using this function

gchamb replied:
since its only console loggin the error

Muaad2001 committed:
perf(client): 🎨 Error messages display user friendly messages and if google or microsoft pop up is closed modal isn't closed

Muaad2001 committed:
fix(client): 🐛 Sign in and create account drop down closes after signin is successful

Muaad2001 committed:
refactor(client):

Muaad2001 committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into Muaad-Signout

nstringham committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into Muaad-Signout

nstringham commented on client/src/components/LobbyChat.svelte:
```diff
@@ -0,0 +1,70 @@
+<script lang="ts">
+  import Modal from "./Modal.svelte";
+  import { onMount, onDestroy } from "svelte";
+  import { authStore } from "$stores/auth";
+  import { onSnapshot, orderBy, query, QueryDocumentSnapshot } from "firebase/firestore";
+  import type { User } from "firebase/auth";
```
these import are not used

you should do `npm run lint` in the cliend directory and try to fix the warnings if they are related to code you wrote

nstringham commented on client/src/components/LobbyChat.svelte:
```diff
@@ -0,0 +1,70 @@
+<script lang="ts">
+  import Modal from "./Modal.svelte";
+  import { onMount, onDestroy } from "svelte";
+  import { authStore } from "$stores/auth";
+  import { onSnapshot, orderBy, query, QueryDocumentSnapshot } from "firebase/firestore";
+  import type { User } from "firebase/auth";
```
these import are not used

you should do `npm run lint` in the cliend directory and try to fix the warnings if they are related to code you wrote

Muaad2001 replied:
resolved

nstringham commented on client/src/components/SigninButton.svelte:
```diff
@@ -1,19 +1,181 @@
+  let errorMessage: string = "";
+  function clearFields() {
+    email = "";
+    password = "";
```
should this also clear `confirmPass`?

Muaad2001 replied:
resolved

nstringham reviewed (changes requested):
really close just a few warning to look at

Muaad2001 committed:
fix(client): 🐛 Confirm password now gets reset when user chooses to not create account and sign in.

nstringham committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into Muaad-Signout

nstringham reviewed (dismissed):
just merged from development everything looks good

nstringham committed:
test(client): ✅ tell playwright to search for full name of lobby buttons

this is because playwright was confusing "Create Lobby" with "Create Account"

zaverbac reviewed (commented):
Everything looks to be fine on my end
